swift/extractor/infra/TargetFile.h

```diff
@@ -2,28 +2,29 @@
 
 #include <string>
 #include <fstream>
-#include <iostream>
-#include <cstdio>
+#include <optional>
 #include <cerrno>
-#include <system_error>
-#include <sstream>
 
 namespace codeql {
 
-// Only the first process trying to open an `TargetFile` for a given `target` is allowed to do
-// so, all others will have an instance with `good() == false` and failing on any other operation.
+// Only the first process trying to create a `TargetFile` for a given `target` is allowed to do
+// so, all others will have `create` return `std::nullopt`.
 // The content streamed to the `TargetFile` is written to `workingDir/target`, and is moved onto
-// `targetDir/target` only when `commit()` is called
+// `targetDir/target` on destruction.
 class TargetFile {
   std::string workingPath;
   std::string targetPath;
   std::ofstream out;
 
  public:
-  TargetFile(std::string_view target, std::string_view targetDir, std::string_view workingDir);
+  static std::optional<TargetFile> create(std::string_view target,
+                                          std::string_view targetDir,
+                                          std::string_view workingDir);
 
-  bool good() const;
-  void commit();
+  ~TargetFile() { commit(); }
+
+  TargetFile(TargetFile&& other) = default;
+  TargetFile& operator=(TargetFile&& other);
```

Comment:
Oof, I hadn't considered we'd need all this.

My understanding is that the clever move assignment operator never gets used. If that's correct, can we delete it so we don't have to think about the non-trivial implementation? That's the approach used by [`gsl::final_action`](https://github.com/microsoft/GSL/blob/d9fc52e20e32fcc804502480a781120b210afd41/include/gsl/util#L74). Having a move constructor but not move assignment is weird, but if we add a comment that it wasn't added simply because it's unused, I think that should be okay.

(Incidentally, using `gsl::final_action` would have made the destructor trivial too, but I don't think it's worth switching to it.)

Reply:
yeah, I was considering it, weighing in the rule of five and whatnot, if you agree I'm happy to delete the move assignment, as I was almost about to